x/authz/keeper/keeper.go

```
@@ -0,0 +1,191 @@
+	return Keeper{
+		storeKey: storeKey,
+		cdc:      cdc,
+		router:   router,
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// getAuthorizationGrant returns grant between granter and grantee for the given msg type
+func (k Keeper) getAuthorizationGrant(ctx sdk.Context, actor []byte) (grant types.AuthorizationGrant, found bool) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(actor)
+	if bz == nil {
+		return grant, false
+	}
+	k.cdc.MustUnmarshalBinaryBare(bz, &grant)
+	return grant, true
+}
+
+func (k Keeper) update(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, updated types.Authorization) error {
+	actor := types.GetActorAuthorizationKey(grantee, granter, updated.MethodName())
+	grant, found := k.getAuthorizationGrant(ctx, actor)
+	if !found {
+		return fmt.Errorf("no authorization found")
+	}
+
+	msg, ok := updated.(proto.Message)
+	if !ok {
+		return fmt.Errorf("cannot proto marshal %T", updated)
+	}
+
+	any, err := codectypes.NewAnyWithValue(msg)
+	if err != nil {
+		return err
+	}
+
+	grant.Authorization = any
+	store := ctx.KVStore(k.storeKey)
+	store.Set(actor, k.cdc.MustMarshalBinaryBare(&grant))
+	return nil
+}
+
+// DispatchActions attempts to execute the provided messages via authorization
+// grants from the message signer to the grantee.
+func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, serviceMsgs []sdk.ServiceMsg) (*sdk.Result, error) {
+	var msgResult *sdk.Result
+	var err error
+	for _, serviceMsg := range serviceMsgs {
+		signers := serviceMsg.GetSigners()
+		if len(signers) != 1 {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "authorization can be given to msg with only one signer")
+		}
+		granter := signers[0]
+		if !bytes.Equal(granter, grantee) {
```

I would prefer to return an error if `granter == grantee`